### PR TITLE
Expand `work_queue` table to accommodate work pools

### PR DESCRIPTION
**Pull request by cicdw**
## Overview
This PR changes the approach for introducing work pools. Instead of creating a new `work_pool_queue` table for work queues with a work pool, this PR extends the existing `work_queue` table to add a `work_pool_id` column and a `priority` column. To accommodate existing work queues, a default work pool is created and all existing work queues are placed under that default work pool. All existing work queue API routes will use the default work pool as well.

#### Schema updates
- Adds `work_pool_id` and `priority` columns to the `work_queue` table
- Drops `work_pool_queue` table
- Updates references to the `work_pool_queue` table to point to `work_queue` table instead

### API
- All routes signatures are unchanged
- Current work queue routes that operate on work queue name will work as before, until the default work pool is created. After the default work pool is created, any routes using only the work queue name will be scoped to the default work pool.

### Client
- Renaming instances of `work_pool_queue` caused some conflicts. Optional work pool parameters we added to the existing work queue `OrionClient` methods and branching logic was introduced to change which route is called depending on if work pool information is provided.

### Schemas
- Removes `WorkPoolQueue` and associated schemas
- Expands `WorkQueue` schema to include work pool and priority information

Subsequent migrations can be seen in https://github.com/PrefectHQ/prefect/pull/8327 and https://github.com/PrefectHQ/prefect/pull/8331

### CODE REVIEW

File: src/prefect/agent.py [hunk 1]
```diff
@@ -141,8 +141,8 @@ async def get_work_queues(self) -> AsyncIterator[Union[WorkQueue, WorkPoolQueue]
         for name in self.work_queues:
             try:
                 if self.work_pool_name:
```
Comment (desertaxle) [inline src/prefect/agent.py:143]: Can remove this if statement

File: src/prefect/agent.py [hunk 2]
```diff
@@ -154,11 +154,8 @@ async def get_work_queues(self) -> AsyncIterator[Union[WorkQueue, WorkPoolQueue]
                     # queues using a regex
                     try:
                         if self.work_pool_name:
```
Comment (desertaxle) [inline src/prefect/agent.py:156]: Can remove this if statement

File: src/prefect/agent.py [hunk 3]
```diff
@@ -214,11 +211,13 @@ async def get_and_submit_flow_runs(self) -> List[FlowRun]:
 
                 else:
                     try:
-                        if isinstance(work_queue, WorkPoolQueue):
-                            responses = await self.client.get_scheduled_flow_runs_for_work_pool_queues(
-                                work_pool_name=self.work_pool_name,
-                                work_pool_queue_names=[work_queue.name],
-                                scheduled_before=before,
+                        if self.work_pool_name:
```
Comment (desertaxle) [inline src/prefect/agent.py:214]: Can remove this if block

**Review by peytonrunyan — APPROVED** [review]
Went over this in a sync review just now. With the understanding that I'm not to be trusted with reviewing migration code, LGTM!

File: src/prefect/orion/database/migrations/versions/postgresql/2023_01_26_152801_0a1250a5aa25_expand_work_queue_table.py [hunk 4]
```diff
@@ -0,0 +1,365 @@
...
+        ).params({"default_pool_id": default_pool_id}),
+    )
+
+    default_queue_id = connection.execute(
+        sa.text(
+            "SELECT id FROM work_queue WHERE name = 'default' and work_pool_id = :default_pool_id"
+        ).params({"default_pool_id": default_pool_id}),
+    ).fetchone()[0]
+
+    connection.execute(
```
Comment (anticorrelator) [inline src/prefect/orion/database/migrations/versions/postgresql/2023_01_26_152801_0a1250a5aa25_expand_work_queue_table.py:161]: I'd break up this data migration into a separate migration, this way we can test to make sure we can do the migration separately in the background

File: src/prefect/orion/database/migrations/versions/postgresql/2023_01_26_152801_0a1250a5aa25_expand_work_queue_table.py [hunk 5]
```diff
@@ -0,0 +1,365 @@
...
+                        WHERE flow_run.id in (
+                            SELECT id 
+                            FROM flow_run 
+                            WHERE flow_run.work_queue_id IS NULL and flow_run.work_queue_name=:name 
+                            LIMIT :batch_size
+                        )
+                        """
+                    ).params({"id": row[0], "name": row[1], "batch_size": batch_size}),
+                )
+                if result.rowcount <= 0:
```
Comment (anticorrelator) [inline src/prefect/orion/database/migrations/versions/postgresql/2023_01_26_152801_0a1250a5aa25_expand_work_queue_table.py:201]: I've found using `result.rowcount < batch_size` to be a more robust break condition; I've run into some problems using `<= 0` before

**Review by anticorrelator — COMMENTED** [review]
Because we have a large migration, I think we should probably break up the schema and data migrations here, as well as separately test that all tests will pass with just the schema migration

File: src/prefect/orion/database/migrations/versions/postgresql/2023_01_26_152801_0a1250a5aa25_expand_work_queue_table.py (hunk 5 above)
Comment (desertaxle) [reply src/prefect/orion/database/migrations/versions/postgresql/2023_01_26_152801_0a1250a5aa25_expand_work_queue_table.py]: Good call! I implemented this in https://github.com/PrefectHQ/prefect/pull/8327

File: src/prefect/orion/database/migrations/versions/postgresql/2023_01_26_152801_0a1250a5aa25_expand_work_queue_table.py (hunk 4 above)
Comment (desertaxle) [reply src/prefect/orion/database/migrations/versions/postgresql/2023_01_26_152801_0a1250a5aa25_expand_work_queue_table.py]: I've reduced the scope of this migration to only be schema changes. The data migration is in https://github.com/PrefectHQ/prefect/pull/8327 and the clean up in is #8331

File: src/prefect/orion/database/migrations/versions/postgresql/2023_01_26_152801_0a1250a5aa25_expand_work_queue_table.py [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+
+    with op.batch_alter_table("work_queue", schema=None) as batch_op:
+        batch_op.create_foreign_key(
+            batch_op.f("fk_work_queue__work_pool_id__work_pool"),
+            "work_pool",
+            ["work_pool_id"],
+            ["id"],
+            ondelete="cascade",
+        )
+        # batch_op.drop_constraint("uq_work_queue__name")
```
Comment (anticorrelator) [inline src/prefect/orion/database/migrations/versions/postgresql/2023_01_26_152801_0a1250a5aa25_expand_work_queue_table.py:105]: did you mean to delete these?

**Review by anticorrelator — APPROVED** [review]
if tests pass, the schema migration looks good to me—is there a test that checks that unmigrated data is backwards compatible somewhere?

**Comment by desertaxle** [comment]
> if tests pass, the schema migration looks good to me—is there a test that checks that unmigrated data is backwards compatible somewhere?

There isn't an explicit test for that, but the existing work queue tests act as a test for backwards compatibility.

**Pull request merged**